File: Implementation of DSU/dsu.java
class dsu {
    private int[] parent;
    private int[] size;

    dsu(int[] parent) {
        this.parent = parent;
        size = new int[parent.length];
        for (int i = 0; i < parent.length; i++) {
            size[i] = 1;
        }
    }

    public void union(int x, int y) {
        int xRoot = find(x);
        int yRoot = find(y);

        if (xRoot == yRoot) return;

        if (size[xRoot] > size[yRoot]) {
            parent[yRoot] = xRoot;
            size[xRoot] += size[yRoot];
        } else {
            parent[xRoot] = yRoot;
            size[yRoot] += size[xRoot];
        }
    }

    public int find(int x) {
        if (parent[x] == x) return x;

        parent[x] = find(parent[x]);
        return parent[x];
    }

    public static void main(String[] args) {
        int[] parent = {0, 1, 2, 3, 4, 5};

        dsu obj = new dsu(parent);
        obj.union(0, 1);
        obj.union(2, 3);
        obj.union(4, 5);

        for (int n : parent) {
            System.out.print(n + " ");
        }
    }
}

// tc => o(log * n)

File: Implementation of DSU/dsu.js
function UNION(x, y, parent, size){
    // parent is the dsu array

    let xRoot = find(x, parent)  // find parent of x
    let yRoot = find(y, parent)
    
    if(xRoot == yRoot) return; // x and y both have same parent

    if(size[xRoot] > size[yRoot]){
        // x group is bigger
        parent[yRoot] = xRoot;
        size[xRoot] += size[yRoot];
    }else{
        // if y has bigger group
        parent[xRoot] = yRoot;
        size[yRoot] += size[xRoot];
    }
}

function find(x, parent){
    if(parent[x] === x) return x // given x is parent
    parent[x] = find(x);  // for path compression attach

    return parent[x];
}

let parent = [0, 1, 2, 3, 4, 5];
let size = [1, 1, 1, 1, 1, 1];

UNION(0, 1, parent, size);
UNION(2, 3, parent, size);
UNION(4, 5, parent, size);

console.log(parent); // Output: [1, 1, 3, 3, 5, 5]
console.log(size); // Output: [1, 2, 1, 2, 1, 2]

// tc => o(log * n)